### ISSUE TITLE
@types/lodash/without wrong type? [newb]

### ISSUE THREAD

**Issue by ehahn9** [DefinitelyTyped/DefinitelyTyped#25916]
@andy-ms @aj-r 

ts newb here - apologies, but:

Seems like `without(Array<T>, undefined)` is typed as `Array<T | undefined>` but should be T since we're removing things, never adding them:

```
const foo = ['abc', 'def'];
without(foo, 'missing'); // GOOD: => string[]
without(foo, undefined); // BAD?: => (string | undefined)[]
```

I'm sure I'm missing something, but maybe the definition should be more like:

```
        without<T>(
            array: List<T> | null | undefined,
            ...values: any[] // *** was T[]
        ): T[];
```

**Comment by aj-r** [comment]
Interesting. We're trying to prevent people from passing a value that couldn't possibly be in the array. For example this:

```ts
const foo = ['abc', 'def'];
without(foo, 1);
```

While this is technically allowed by lodash, you should never do it, and it often indicates a bug in your code. That's why we want to avoid using `any[]`. Your example falls in the same category:

```ts
const foo = ['abc', 'def'];
without(foo, undefined);
```

We really want a compiler error here (`undefined is not assignable to type string`), but instead the compiler infers `T` to be `string|undefined` so you get the bad return type here. I'd say that this is the best "expected behavior" we can get right now - the fix is to make sure you don't pass `undefined` to `without`.

One could make an argument that you should be able to pass `string|undefined` values to `without`. The issue then is, where do we stop with what we allow? By that logic, something like `string|number` should also be allowed, but I don't think it should be. Instead, there are a number of solutions to the `string|undefined` use case that don't require us to loosen the type definitions:

```ts
const foo = ['abc', 'def'];
declare const value: string | undefined;
if (value !== undefined)
    without(foo, value);
without(foo, value || "");
without(foo, value!);
`

**Comment by ehahn9** [comment]
tl;dr: thanks - makes sense (closing).

Ahh, I see (and agree). Thanks! In my particular case, it was something like this:

```
interface Things<T> {
  current?: T;
  history: T[]
}

// contrived usage:
history = without(history, current);
```

Which seemed harmless. I briefly went down the hackity-hack path (`without=history(without, current!)` but I see now that better (clearer for sure!) would be with an explicit guard:

```if (current) history = without(history, current)```

So your rationale did exactly what you wanted: discourage the use of types not in the collection.

Thanks for the explanation.
